profondo/cnn.py

```diff
@@ -86,6 +95,96 @@ def build_model(x_train, y_train):
+        # Build the "bottleneck-features" to feed into the top layers, to init
+        # the output layers with reasonable features: we basically record the
+        # output of the last convolutional block with our data and feed that
+        # into the new top block.
+        # Note: our images are already scaled in the [0, 1] range, we don't need
+        # the "rescale" part as required by the pre-trained networks. We
+        # don't shuffle the samples to be able to pick their labels later on.
+        init_batch = 100
+        num_train_batches = len(x_train) // init_batch
+        bottleneck_train = model.predict(x_train, num_train_batches, verbose=True)
+        num_val_batches = len(x_val) // init_batch
+        bottleneck_val = model.predict(x_val, num_val_batches, verbose=True)
+
+        genre_weights = compute_genre_weights(y_train)
+
+        # Compile and train the top_model with the "bottleneck-features".
+        top_model = generate_top_model(bottleneck_train.shape[1:], y_train.shape[1])
+        top_model.compile(loss=keras.losses.binary_crossentropy,
+                          optimizer=keras.optimizers.SGD(lr=0.01, momentum=0.9))
+        logging.info(top_model.summary())
+        callbacks = [
+            ReduceLROnPlateau(monitor='val_loss', factor=0.2, patience=5, min_lr=10e-6),
+            TensorBoard(log_dir=join(TB_DIR, 'top'), histogram_freq=1, write_graph=True, write_images=False)
+        ]
+        top_model.fit(bottleneck_train, y_train,
+                      epochs=50,
+                      batch_size=init_batch,
+                      validation_data=(bottleneck_val, y_val),
+                      class_weight=genre_weights,
+                      callbacks=callbacks,
+                      verbose=True)
+
+        # Save the weights for top_model. They will be used to init a new instance of
+        # the model that will be linked to the convolutional layers.
+        top_model.save_weights(TOP_MODEL_WEIGHTS_FILE)
+    
+    # Instantiate another top_model. We need to do the weights save/load dance
+    # as we cannot modify a model after it was compiled, so we cannot use the
+    # previous |top_model| again.
+    trained_top = generate_top_model(model.output_shape[1:], y_train.shape[1])
+    trained_top.load_weights(TOP_MODEL_WEIGHTS_FILE)
+
+    # Set all the convolutional layer in the pre-trained network to
+    # to non-trainable (weights will not be updated).
+    for layer in model.layers:
+        layer.trainable = False
+
+    # Join the convolutional part and the top layers. See
+    # https://github.com/fchollet/keras/issues/4040#issuecomment-253309615
+    joint_model = Model(input=model.input, output=trained_top(model.output))
+
+    # compile the model with a SGD/momentum optimizer
+    # and a very slow learning rate.
+    joint_model.compile(loss=keras.losses.binary_crossentropy,
+                        optimizer=keras.optimizers.SGD(lr=1e-4, momentum=0.9))
```

Comment:
It seems to work better than Adam when using a pre-trained network on my system (Adam diverges, while SGD seem to converge to a nicer optimum)